### PR TITLE
Rework local storage for node table and fix #2376

### LINKED ISSUE #2376: Cannot edit or delete node... after editing once
Version: 0.0.12

### Cannot edit twice
Below code should change the node name twice... resulting in "rename 2x", but it gets stuck at "rename 1x"
Even if I reset connection, and immediately run "rename 2x"... it will never change after being edited once.

```
import kuzu
import time

db = kuzu.Database("testdb")
conn = kuzu.Connection(db)

#create tables
conn.execute("CREATE NODE TABLE thing(name STRING, PRIMARY KEY (name))")
conn.execute("MATCH (n) DELETE n")

#create node
conn.execute("MERGE (n:thing {name: 'original'})")

print("rename 'original' --> 'rename 1x'")
results = conn.execute("MATCH (n:thing {name: 'original'}) SET n.name = 'rename 1x' RETURN n")
while results.has_next():
    print(results.get_next())

time.sleep(1) #wait... to prove it's not waiting for commit

print("rename 'rename 1x' --> 'rename 2x'")
results = conn.execute("MATCH (n:thing {name: 'rename 1x'}) SET n.name = 'rename 2x' RETURN n")
while results.has_next():
    print(results.get_next())

print("print nodes...")
results = conn.execute("MATCH (n) RETURN n")
while results.has_next():
    print(results.get_next())
```

Output:
```
rename 'original' --> 'rename 1x'
rename 'rename 1x' --> 'rename 2x'
print nodes...
[{'_label': 'thing', '_id': {'offset': 0, 'table': 0}, 'name': 'rename 1x'}]
```

### Cannot delete after edit
Also.... I cannot delete the node after it's been edited once.  
I.e. ... after running conn.execute("MERGE (n:thing {name: 'original'})")

This works
```
conn.execute("MATCH (n:thing) WHERE n.name='original' DELETE n")
```

This does NOT work
```
conn.execute("MATCH (n:thing {name: 'original'}) SET n.name = 'rename 1x' RETURN n")
conn.execute("MATCH (n:thing) WHERE n.name='rename 1x' DELETE n")
```

### PR DESCRIPTION
**Pull request by ray6080**
- Rework local storage to be more space efficient. Instead of always allocating a Vector of 2048 whenever an update/insert resides in its offset range. For example, update at offset 0 and offset 2049 will allocate two Vectors. In this PR, updates/inserts always append to the end of a collection of Vectors, which allocates as needed and one value at a time. The mapping between node offsets and row idx is maintained in extra maps.
- Fix issue 2376. The issue is due to incorrectly updating hash index when updates on node table involves the primary key column.

Note that this PR also forces flattening on updates/deletions/insertions, which is to simplify the implementation of following fixes on rel updates/deletions/insertions. This can bring performance down, though it's not profiled by how much, and should be solved at once as we optimize the update performance after all fixes.
Also, there are quite few places not covered by tests, which are mostly due to that lots of tests on rel updates are disabled. I will revisit the test coverage after my fixes on rel updates.

**TODOs**
- [ ] Revisit and improve storage test coverage.
- [ ] Remove the rigid flatten all rules on updates/insertions/deletions.

### CODE REVIEW

**Review by acquamarin — APPROVED** [review]

File: src/include/storage/local_storage.h [hunk 1]
```diff
@@ -4,36 +4,48 @@
...
+    // Note: `initializeLocalTable` should be called before `insert` and `update`.
+    void initializeLocalTable(
+        common::table_id_t tableID, const std::vector<std::unique_ptr<Column>>& columns);
+    void insert(common::table_id_t tableID, common::ValueVector* nodeIDVector,
+        const std::vector<common::ValueVector*>& propertyVectors);
+    void update(common::table_id_t tableID, common::ValueVector* nodeIDVector,
+        common::column_id_t columnID, common::ValueVector* propertyVector);
+    void delete_(common::table_id_t tableID, common::ValueVector* nodeIDVector);
+
+    inline std::unordered_set<common::table_id_t> getTablesToUpdate() {
```
[the remark this answers is not in the document]
Comment (ray6080) [reply src/include/storage/local_storage.h:34]: I think we can just adopt the convention of `delete_` unless there is serious concern over this, which I don't have.

File: src/common/vector/value_vector.cpp [hunk 2]
```diff
@@ -451,6 +451,7 @@ void StringVector::addString(ValueVector* vector, ku_string_t& dstStr, ku_string
     if (ku_string_t::isShortString(srcStr.len)) {
         dstStr.setShortString(srcStr);
     } else {
+
```
[the remark this answers is not in the document]
Comment (ray6080) [reply src/common/vector/value_vector.cpp:454]: Make sense, renamed to `getTableIDsWithUpdate`.

File: src/storage/store/struct_column.cpp [hunk 3]
```diff
@@ -24,7 +25,22 @@ StructColumn::StructColumn(LogicalType dataType, const MetadataDAHInfo& metaDAHe
...
+void StructColumn::scan(node_group_idx_t nodeGroupIdx, ColumnChunk* columnChunk) {
+    KU_ASSERT(columnChunk->getDataType().getPhysicalType() == PhysicalTypeID::STRUCT);
+    nullColumn->scan(nodeGroupIdx, columnChunk->getNullChunk());
+    if (nodeGroupIdx >= metadataDA->getNumElements()) {
+        columnChunk->setNumValues(0);
+    } else {
+        auto chunkMetadata = metadataDA->get(nodeGroupIdx, TransactionType::WRITE);
+        columnChunk->setNumValues(chunkMetadata.numValues);
+    }
+    auto structColumnChunk = ku_dynamic_cast<ColumnChunk*, StructColumnChunk*>(columnChunk);
```
[the remark this answers is not in the document]
Comment (ray6080) [reply src/storage/store/struct_column.cpp:37]: hmm. I'm indifferent. I can change this in a separate PR.

File: src/include/storage/local_table.h [hunk 4]
```diff
@@ -1,169 +1,148 @@
...
-        const common::LogicalType& logicalType, MemoryManager* mm);
-};
+    LocalVectorCollection(const common::LogicalType& dataType, MemoryManager* mm)
+        : dataType{dataType}, mm{mm}, numRows{0} {}
+
+    void read(common::row_idx_t rowIdx, common::ValueVector* outputVector,
+        common::sel_t posInOutputVector);
+    void insert(common::ValueVector* nodeIDVector, common::ValueVector* propertyVectors);
+    void update(common::ValueVector* nodeIDVector, common::ValueVector* propertyVector);
+    inline void delete_(common::offset_t nodeOffset) {
```
[the remark this answers is not in the document]
Comment (ray6080) [reply src/include/storage/local_table.h:47]: I think we can just adopt the convention of delete_ unless there is serious concern over this, which I don't have.

File: src/include/storage/local_storage.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (ray6080) [reply src/include/storage/local_storage.h]: Make sense, renamed to getTableIDsWithUpdate.

**Pull request merged**